### PR TITLE
Add identity to task execution metadata

### PR DESCRIPTION
**Pull request by noahjax**
## Why are the changes needed?
For our specific use case it is critical to be able to determine the identity for the subject who launches a task. It seems like other information on the security context could also be more generally useful, so I included that in addition to the critical `execution_identity` field as part of this PR.

## What changes were proposed in this pull request?
Add `identity` to `TaskExecutionMetadata`. In conjunction with https://github.com/flyteorg/flytekit/pull/2282, this change will make identity available to flyte agents.

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] All commits are signed-off.

## Related PRs
https://github.com/flyteorg/flyte/issues/3936

### CODE REVIEW

File: flyteplugins/go/tasks/plugins/webapi/agent/plugin.go [hunk 1]
```diff
@@ -347,13 +347,16 @@ func getFinalAgent(taskCategory *admin.TaskCategory, cfg *Config, agentRegistry
 
 func buildTaskExecutionMetadata(taskExecutionMetadata core.TaskExecutionMetadata) admin.TaskExecutionMetadata {
 	taskExecutionID := taskExecutionMetadata.GetTaskExecutionID().GetID()
+	securityContext := taskExecutionMetadata.GetSecurityContext()
+
 	return admin.TaskExecutionMetadata{
 		TaskExecutionId:      &taskExecutionID,
 		Namespace:            taskExecutionMetadata.GetNamespace(),
 		Labels:               taskExecutionMetadata.GetLabels(),
```
[the remark this answers is not in the document]
Comment (noahjax) [reply flyteplugins/go/tasks/plugins/webapi/agent/plugin.go:355]: Yeah I think those labels may only apply when creating k8s pods. This is what I see in the labels when I log them
```
{"json":{"exec_id":"a5jknwxdqf5jbphrbp2b","node":"Permagenttest2","ns":"domino-compute","res_ver":"21567793","routine":"worker-3","src":"plugin.go:105","tasktype":"domino_job","wf":"65ca8301c1052944abab8249:development:.flytegen.Perm agent test 2"},"level":"warning","msg":"Labels: node-id=\"permagenttest2\"\ntask-name=\"perm-agent-test-2\"\ninterruptible=\"false\"\ndomain=\"development\"\nexecution-id=\"a5jknwxdqf5jbphrbp2b\"\nproject=\"65ca8301c1052944abab8249\"\nshard-key=\"5\"\nworkflow-name=\"flytegen-perm-agent-test-2\"\n","ts":"2024-03-26T03:45:24Z"}
```

[the remark this answers is not in the document]
Comment (pingsutw) [reply flyteplugins/go/tasks/plugins/webapi/agent/plugin.go:355]: that label can be applied to the non-k8s tasks. For example, AWS batch, spark on databricks, etc.

**Review by pingsutw — DISMISSED** [review]

**Comment by pingsutw** [comment]
@noahjax just curious, so you are going to use [OAuth2TokenRequest](https://github.com/flyteorg/flyte/blob/97a79c030555eaefa3e27383d9b933ba1fdc1140/flyteidl/protos/flyteidl/core/security.proto#L129) in the agent not `Secret`, right? 

My PR is going to pass the actual secret value to the agent because we save the secret in AWS/GCP secret manager, and only the propeller has permission to read it.

**Comment by noahjax** [comment]
@pingsutw For my use case I actually only need the `execution_identity`, but I wanted to grab that in a way that could be useful in other cases as well. Your PR is probably more useful in general because it actually fetches secret values, but it would be awesome if in addition to those secret values we could also get some information from the `run_as` field to the agent. That might require some new message type instead of piggy-backing in the existing `SecurityContext`

**Comment by pingsutw** [comment]
@noahjax In that case, should we just pass [Identity](https://github.com/flyteorg/flyte/blob/97a79c030555eaefa3e27383d9b933ba1fdc1140/flyteidl/protos/flyteidl/core/security.proto#L117C5-L117C13) to the agent? flyte won't mount the `Secret` and `OAuth2TokenRequest` to the agent, since it's a long-running service

**Comment by noahjax** [comment]
@kumare3 @pingsutw That makes sense, I updated the PR to use `Identity` instead of `SecurityContext`

**Review by pingsutw — APPROVED** [review]

**Comment by pingsutw** [comment]
cc @ddl-ebrown @noahjax we can merge it, right

**Comment by noahjax** [comment]
Yeah go for it

File: flyteplugins/go/tasks/plugins/webapi/agent/plugin.go [hunk 2]
```diff
@@ -347,13 +347,15 @@ func getFinalAgent(taskCategory *admin.TaskCategory, cfg *Config, agentRegistry
...
 	taskExecutionID := taskExecutionMetadata.GetTaskExecutionID().GetID()
+
 	return admin.TaskExecutionMetadata{
 		TaskExecutionId:      &taskExecutionID,
 		Namespace:            taskExecutionMetadata.GetNamespace(),
 		Labels:               taskExecutionMetadata.GetLabels(),
 		Annotations:          taskExecutionMetadata.GetAnnotations(),
 		K8SServiceAccount:    taskExecutionMetadata.GetK8sServiceAccount(),
 		EnvironmentVariables: taskExecutionMetadata.GetEnvironmentVariables(),
+		Identity:             taskExecutionMetadata.GetSecurityContext().RunAs,
```
[the remark this answers is not in the document]
Comment (eapolinario) [reply flyteplugins/go/tasks/plugins/webapi/agent/plugin.go:358]: cc: @pingsutw

[the remark this answers is not in the document]
Comment (pingsutw) [reply flyteplugins/go/tasks/plugins/webapi/agent/plugin.go:358]: @ddl-ebrown would you create another PR to address it?

**Pull request merged**